### ISSUE TITLE
proposal: reflect: add `Value.CallInto`

### ISSUE THREAD

**Issue by twmb** [golang/go#49340]
`reflect.Call` allocates a `[]reflect.Value` slice of length `nout` for the number of return arguments from the function. This is fine for a one-off call of a function, but when repeatedly invoking a function, `reflect.Call` allocations can quickly add up.

As an example, I have a helper package that sorts arbitrary packages, and if a type has a `func (t T) Less(other T) bool` (i.e., `Name == "Less", NumIn() == 1, NumOut() == 1, In(0) == T, Out(0) == bool`), the package calls that function for sorting. This is done within sort.Slice, meaning `Less` is called as many times as `sort.Slice` needs to compare. For an ordered slice of 1000 elements, each `sort.Slice` allocates 25735 times. 

The only four allocations in this benchmark are (tip @ 2cf85b1fb8)
```
  105.50MB 68.28% 68.28%   105.50MB 68.28%  reflect.Value.call /home/twmb/go/go.tip/src/reflect/value.go:565
      41MB 26.53% 94.81%       41MB 26.53%  reflect.methodReceiver /home/twmb/go/go.tip/src/reflect/value.go:862
    4.50MB  2.91% 97.72%     4.50MB  2.91%  reflect.Value.call /home/twmb/go/go.tip/src/reflect/value.go:609
    1.50MB  0.97% 98.70%     1.50MB  0.97%  runtime.allocm /home/twmb/go/go.tip/src/runtime/proc.go:1866
```

70% of the allocations are from allocating the output argument slice. If possible, I could provide the input slice and reuse it for all `Call`s, which would effectively eliminate this alloc.

I looked into line 862 and into the runtime, I'm not sure how to get rid of that allocation, the code I see doesn't obviously look to allocate; if possible it would be quite nice to get rid of this alloc as well. Line 609 has a TODO that could eliminate the allocation. runtime/proc.go:1866 looks to be unrelated.

`reflect.Value.call` can be:
```go
func (v Value) call(op string, in, out []Value) []Value {
```

If out is non-nil, then it is used rather than allocating a new slice. If the input `out` is too small, the code will panic. Essentially, line 565 turns into:
```go
ret = out
if ret == nil {
    ret = make([]Value, nout)
}
if len(ret) < nout { // optional check
    panic("short output slice")
}
```

**Comment by ianlancetaylor** [comment]
Going forward would it make sense to write your package using generics?

**Comment by twmb** [comment]
I don't think so. This is doing deep sorting on struct fields (admittedly this is not a high priority package). If the field is a slice, I check the inner type for the method. With generics, even if I know the method has a `Less` function, I cannot type assert into `Lesser[T]` because T requires instantiation and I do not know the type.

**Comment by rsc** [comment]
If we add CallInto or something like it, does reflect.Call go down to zero allocations?
It would be good to know we were done with this particular issue, at least for Call.

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://golang.org/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by dsnet** [comment]
This seems like a duplicate of #43732.

> If we add CallInto or something like it, does reflect.Call go down to zero allocations?

According to my comment https://github.com/golang/go/issues/43732#issuecomment-761704581, we can get it down to zero allocations.

**Comment by twmb** [comment]
Interesting, yes I think this is a duplicate of #43732. Does the CL linked take care of methods? I don't see a change in `methodReceiver`, which was the remaining allocation I ran into.

I also tried removing the allocation locally -- I got a similar sort of speedup: not as much as I hoped, but the allocation drop is somewhat nice regardless.

**Comment by dsnet** [comment]
> Does the CL linked take care of methods? 

I didn't do anything special for method functions. I'll take a look later if something is needed there.

**Comment by rsc** [comment]
Indeed, this is a duplicate of #43732, but this one is in the active column, so I'll close that one.

**Comment by rsc** [comment]
It sounds like people basically agree about the signature (even in #43732): it takes two []Value and returns nothing.
But we need a name, and we need to figure out whether it handles variadic functions like Call or like CallSlice.
If it handles them like Call, doesn't that allocate? So probably it should handle them like CallSlice.
Maybe it would be CallSlices?
Other names welcome but CallInto and CallWith both sound like they are missing something.

**Comment by rsc** [comment]
What should we do about Call vs CallSlice? We probably don't want to add two new functions.
And if we do this, are we down to zero allocations, so we know we won't need to add another variant in the future?

**Comment by Splizard** [comment]
The naming problem could be avoided by introducing a `reflect.Caller` type.
```go
//Caller is a low-level type used to make 
//zero-allocation function calls.
type Caller struct {}
```
I suspect that adding this type has an additional advantage, it leaves the door open to a `NewCaller/Caller` constructor of the  type being able to optimise repeat calls to the function it's calling. 
```go
echo := reflect.ValueOf(func(in string) string { return in }).Caller()
```
or
```go
echo := reflect.NewCaller(reflect.ValueOf(func(in string) string { return in }))
```
then
```go
results := make([]reflect.Value, 1)
args := []reflect.Value{
    reflect.ValueOf("hello"),
}

for i := 0; i < 100; i++ {
    echo.Call(args, results) //more efficient
}
```

> What should we do about Call vs CallSlice?

`reflect.Caller` provides at least two options:

1. Have two methods on the `reflect.Caller` type, `Call` and `CallSlice`.
2. Have `reflect.Caller.Call` behave like `CallSlice` when it is constructed from a variadic function.

It can then be considered that `reflect.Value.Call` and `reflect.Value.CallSlice` are shorthand/abstractions around the use of the "low-level" Caller type.

> And if we do this, are we down to zero allocations, so we know we won't need to add another variant in the future?

In my `echo` example above, it is my understanding that the string header returned from `echo` will escape to the heap when it is converted into a reflect.Value and placed in the `results` slice. Therefore 'zero-allocation' is only possible by providing pointers in the results slice for Call to write into. I think it should be the requirement of `reflect.Caller`'s user to allocate all the results needed to call the function. Only `reflect.Value.Call`, `reflect.Value.CallSlice` and `reflect.Caller`'s constructor should be allowed to allocate.  

```go
results := []reflect.Value{new(string)}
args := []reflect.Value{reflect.ValueOf("hello")}
for i := 0; i < 100; i++ {
    echo.Call(args, results) //zero-allocations.
}
```

**Comment by twmb** [comment]
Adding the ability to specify the output slice removes allocations for non-method function calls. For methods, there still is an allocation to create a method pointer. Since `call` immediately uses and then discards this method pointer, it would be good if this allocation could be removed, but I'm not sure how, and removing the allocation for output arguments does not remove the method pointer allocation.

**Comment by rsc** [comment]
It sounds like we are having trouble getting to an API that will actually guarantee zero allocations and that is simple enough to be a reasonable API change. Does anyone see how to thread this needle?

**Comment by rsc** [comment]
It still sounds like we don't have a final API to decide about,
one that will guarantee zero allocations and is simple enough to present to users.
Does anyone want to try to do that?

**Comment by Splizard** [comment]
@twmb
Is this something that a `reflect.Caller` type could handle? If the `reflect.Value` is a method, allocate this method pointer and reuse it across sequential calls.

**Comment by twmb** [comment]
I think `reflect.Caller` is a nifty idea: the method pointer is allocated once, and function lookups are cached (and this work eliminated) across repeated invocation. I'm not sure what the code would look like but I'm :+1: on the idea.

Also, `reflect.Caller` _could_ have one single method. Stealing some documentation from the current `reflect.Value.Call`,
```go
// Call calls the function v with the input arguments in. For example, if len(in)
// == 3, v.Call(in) represents the Go call v(in[0], in[1], in[2]).  It returns the
// output results as Values. As in Go, each input argument must be assignable to
// the type of the function's corresponding input parameter. If v is a variadic
// function, Call creates the variadic slice parameter itself, copying in the
// corresponding values.
//
// The returned slice is reused across repeated invocations to call. If you wish
// to keep the argument slice itself, it must be copied.
func (c reflect.Caller) Call(args []reflect.Value) []reflect.Value
```

In this scheme, the returned argument slice can be allocated once on the first call, and reused across future invocations.

Alternatively, this type could have a second method. `CallInto` was questioned above, and I'm not so sure on `CallWith` myself (with what?), but perhaps `CallWithReturns` (still open to better names)?

In summary for new APIs,

```go
func NewCaller(interface{}) reflect.Caller
func (v reflect.Value) Caller() reflect.Caller

type reflect.Caller struct { ... }

func (c reflect.Caller) Call(args []reflect.Value) []reflect.Value // reuses return slice across repeated invocations
func (c reflect.Caller) CallWithReturns(args, returns []reflect.Value)
```

I believe it should be possible to have zero allocation calls with this helper type.

**Comment by rsc** [comment]
Is Caller just a cache for allocated things? If so, maybe it shouldn't embed the receiver? That is, maybe it should be 

var caller reflect.Caller
caller.Call(v, args)

because then you can do

caller.Call(otherV, otherArgs)

Does anyone want to try implementing this?

**Comment by rsc** [comment]
Sounds like we are waiting on someone to implement this to understand whether it satisfies the performance goals for the proposal.

**Comment by rsc** [comment]
**[Placed on hold](https://golang.org/s/proposal-status#hold)**.
— rsc for the proposal review group

**Comment by nrwiersma** [comment]
I have taken a run at implementing this, which can be found here: https://github.com/nrwiersma/go/blob/caller/src/reflect/caller.go 
and it seems it is possible to get it down to zero allocs.

This is basically just a copy of `Value.Call` that caches a bunch of the values, and I am not sure I fully understand everything that happens here. It seems that caching the output of `funcLayout` reduces the time taken by roughly half (~80ns of ~160ns) and from what I gather from its description should be safe. Without caching it the call time is basically unchanged while the allocs are removed. 

I chose, at least for this PoC, not to allow the user to provide their own output slice, as to remove the alloc from L270 a fair amount of type checking was needed, and it was starting to get really ugly. For now the values are created on construction and reused.

My benchmark numbers are as follows:
```
name           time/op
Value_Call-8    159ns ± 1%
Caller_Call-8  76.1ns ± 0%

name           alloc/op
Value_Call-8    32.0B ± 0%
Caller_Call-8   0.00B

name           allocs/op
Value_Call-8     2.00 ± 0%
Caller_Call-8    0.00
```

While I was at it I noticed that `callMethod` is also allocating on `methodReceiver` and takes a long time on `funcLayout`, but this is likely for another issue.

**Comment by ianlancetaylor** [comment]
Thanks. Taking off hold.

**Comment by rsc** [comment]
The implemented API seems to be:

```
package reflect

type Caller struct { ... }
func NewCaller(v Value) *Caller 
func (c *Caller) Call(rcvr Value, in []Value) []Value 
```

This lets NewCaller compute various things about v that can be reused across multiple calls to v.
I'm not sure whether it should be a method on Value instead of a top-level function,
and of course we'd want to deduplicate the code with reflect.Call
(presumably Value.Call would use NewCaller), but it looks reasonable.

@twmb, @dsnet, any thoughts?

**Comment by nrwiersma** [comment]
> (presumably Value.Call would use NewCaller)

NewCaller.Call is using `CallSlice` style input args, so this would be a change to the way that `Value.Call` behaves.

**Comment by rsc** [comment]
@nrwiersma thanks for the reminder about CallSlice. It seems like Caller should have both Call and CallSlice methods.
Then both Value.Call and Value.CallSlice would be wrappers.

@twmb, @dsnet any thoughts?

**Comment by nrwiersma** [comment]
@rsc I have refactored as requested. `Caller` has moved into https://github.com/nrwiersma/go/blob/caller/src/reflect/value.go#L432 and implements both `Call` and `CallSlice`. It is now accessed via `Value.Caller()` and `Value.Call` and `Value.CallSlice` both now use `Caller`. `Value.call` has moved to `Caller.call` and I needed to add `methodReceiverType` to check the actual receiver type.

All tests are passing and my benchmarks look like this:

```
name                      time/op
Call-8                       121ns ± 7%
CallerCall-8                11.7ns ± 1%
CallMethod-8                 141ns ± 0%
CallerCallMethod-8          33.3ns ± 0%

name                      alloc/op
Call-8                        256B ± 0%
CallerCall-8                 0.00B
CallMethod-8                  256B ± 0%
CallerCallMethod-8           0.00B

name                      allocs/op
Call-8                        1.00 ± 0%
CallerCall-8                  0.00
CallMethod-8                  1.00 ± 0%
CallerCallMethod-8            0.00
```

**Comment by twmb** [comment]
+1 to the current API. How is this avoiding allocations on the return slice?

**Comment by rsc** [comment]
So the current API is:

```
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) Call(rcvr Value, in []Value) []Value 
func (c *Caller) CallSlice(rcvr Value, in []Value) []Value 
```

and the current Value.Call and Value.CallSlice become wrappers:

```
func (v Value) Call(in []Value) []Value {
	var rcvr Value
	if v.flag&flagMethod != 0 {
		rcvr = v
	}
	return v.Caller().Call(rcvr, in)
}
```

The detail about the receiver _not_ being pre-bound already in the result of v.Caller() seems wrong. 
v.Caller().Call(in) should be the same as v.Call(in) in all cases.
That means the Call and CallSlice methods should not have a rcvr Value argument.

So the API I'm suggesting would change to:

```
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) Call(in []Value) []Value 
func (c *Caller) CallSlice(in []Value) []Value 
```

and the wrapper would become

```
func (v Value) Call(in []Value) []Value {
	return v.Caller().Call(in)
}
```

Does that seem reasonable?

**Comment by nrwiersma** [comment]
Fair. Will make the change. It does simplify things a little.

**Comment by dsnet** [comment]
My concern with this API is that I don't know how I'm suppose to use it.

I looked at two use-cases to validate this:
* use of `reflect.Value.Call` in "github.com/google/go-cmp"
* hypothetical use of `reflect.Value.Calue` in "encoding/json" for #5901
    * (we no longer need `reflect.Value.Call` for this now that we have Go generics, but it's still worthwhile to consider as a case study)

In order for this to be useful, I'm responsible for pooling the `reflect.Caller` objects. How do I do that when I'm managing an unbounded number of discrete functions? For each function, I can't just have one pre-prepared `reflect.Caller` since `reflect.Caller` is not concurrent safe. I need a pool of them for each function. It does not scale to have a `sync.Pool` for each of these.

**Comment by nrwiersma** [comment]
To deal with the concurrent case, the user would need to provide the output slice, or the output slice just has to allocate.
One option is the add `CallWithOut(in, out []Value)` and `CallSliceWithOut(in, out []Value)` (names can obviously be changed) which would then be concurrency safe while not changing the general non-concurrent use case which would still reuse the output slice. The other option is to always have to provide the output slice.

**Comment by dsnet** [comment]
> One option is the add `CallWithOut(in, out []Value)` and `CallSliceWithOut(in, out []Value)`...

That's essentially what's proposed in #43732 and implemented in https://golang.org/cl/284222.

Speaking of concurrency, I think that's also one of the dangers of the API of:
```go
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) Call(in []Value) []Value 
func (c *Caller) CallSlice(in []Value) []Value 
```
It's not obvious that `Caller.Call` is not concurrent safe. It's a method that takes an input and returns an output and the functionality doesn't seem to have any reason why it mutates `c`. It seems like it serves a similar purpose as `regexp.Regexp`, where the receiver is some statically pre-computed state. Thus, this may be easy to accidentally hold wrong.

**Comment by rsc** [comment]
One benefit of this API that was not present in #43732 is that the Caller implementation can cache lots of information about v that doesn't have to be recomputed (like frame layout offsets). 

@dsnet, what do you think about keeping the Caller type and that cacheability but also add the out slice as in https://github.com/golang/go/issues/49340#issuecomment-1347710586 as a parameter to make it concurrency-safe?

We would probably also need to add:

    func (c *Caller) NumOut() int

so that the user of this code knows how big an out slice to allocate.

Thoughts?

**Comment by nrwiersma** [comment]
@rsc To be clear, are you suggesting adding `out` slices to the `Call` and `CallSlice` methods, or adding `CallWithOut` and `CallSliceWithOut` functions?

**Comment by ianlancetaylor** [comment]
I think the suggestion is to add an `out` parameter to the `Call` and `CallSlice` methods.  There's no reason for additional methods.

**Comment by dsnet** [comment]
> @dsnet, what do you think about keeping the Caller type and that cacheability but also add the out slice as in https://github.com/golang/go/issues/49340#issuecomment-1347710586 as a parameter to make it concurrency-safe?

Generally, that SGTM. So essentially:
```go
func (c *Caller) Call(out, in []Value)
func (c *Caller) CallSlice(out, in []Value)
```
It's unfortunate that `out` and `in` are of the same type, so it will be easy to accidentally swap the order, but that will hopefully be trivially detected by any unit test.

What's the behavior if the provided `out` has elements that mismatch what the function actually outputs? I propose:
* If the n-th output type is assignable to `out[n].Type()` and `out[n].CanSet`, then the output is directly stored using the equivalent of `out[n].Set`.
* Otherwise, the implementation allocates a new value for `out[n]` and overwrites the old Value at `out[n]`.
* After calling `Call`, every `out[n].CanSet` reports true so that we can reuse resources between calls. Note that this goes contrary to the existing `Value.Call` method, which always returns Values that are not settable.

> We would probably also need to add...

If we add `NumOut`, we probably want `NumIn` to be consistent.

**Comment by nrwiersma** [comment]
The changes have been made. The API is now:
```
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) NumIn() int
func (c *Caller NumOut() int
func (c *Caller) Call(in, out []Value)
func (c *Caller) CallSlice(in, out []Value)
```

@dsnet There is no need add complexity by looking at existing values in the output slice, the output values from the function are set to the slice, regardless of what is in there.

**Comment by dsnet** [comment]
> There is no need add complexity by looking at existing values in the output slice, the output values from the function are set to the slice, regardless of what is in there.

Without doing that, I fail to see how this implementation can possibly be concurrency safe and be allocation free. The underlying storage for each output value must exist somewhere.

**Comment by nrwiersma** [comment]
Fair point about the allocation. I had forgotten I had dealt with this in previous code. I have added the code back in that deals with reusing the value. 

Instead of looking for `flagAddr` I am looking for `flagIndir` which seems to have the desired effect without breaking `Value.Call` which would not be good. The check is this:
```
s := out[i].ptr
if out[i].typ != tv || out[i].flag&flagIndir != flagIndir {
	s = unsafe_New(tv.common())
}
```

**Comment by ianlancetaylor** [comment]
I think it should be `out, in []Value` to match the builtin `copy` function.

**Comment by nrwiersma** [comment]
Fair, the API is now:
```
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) NumIn() int
func (c *Caller NumOut() int
func (c *Caller) Call(out, in []Value)
func (c *Caller) CallSlice(out, in []Value)
```

**Comment by dsnet** [comment]
I would have preferred the ordering of `out, in`, but it's seems have prior precedence for `in, out` with the [existence of `FuncOf`](https://pkg.go.dev/reflect#FuncOf):
```go
func FuncOf(in, out []Type, variadic bool) Type
```
There's a stronger correlation between `Caller.Call` and `FuncOf` then there is with `Copy`.

**Comment by jimmyfrasche** [comment]
FuncOf corresponds to declaration and Call with application

**Comment by rsc** [comment]
Talked to @ianlancetaylor and others at proposal review. Let's go back to in, out. Otherwise this seems good. Any objections?

**Comment by rsc** [comment]
The current proposal is:

```
type Caller struct { ... }
func (v Value) Caller() *Caller
func (c *Caller) NumIn() int
func (c *Caller) NumOut() int
func (c *Caller) Call(in, out []Value)
func (c *Caller) CallSlice(in, out []Value)
```

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://go.dev/s/proposal-status#likely-accept)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://go.dev/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group

**Comment by nrwiersma** [comment]
I have run into an issue with reusing the output values. Initially I had used the following to reuse where possible:
```
s := out[i].ptr
if out[i].typ != tv || out[i].flag&flagIndir != flagIndir || s == nil {
	s = unsafe_New(tv.common())
	out[i] = Value{tv.common(), s, flagIndir | flag(tv.Kind())}
}
```
found [here](https://github.com/nrwiersma/go/blob/caller/src/reflect/value.go#L712)

This was to not break the current behaviour of call, but as it turns out, this allows for the following in the `out`:
```
var i int
out := []Value{ValueOf(i)}
```
This cases some serious weirdness, and it seems I have not way to detect this case, as the only flag set is `flagIndir` (outside of kind flags).

This could be resolved by checking `out[i].CanSet()` but then the newly created `Value` must have `flagAddr` which, as stated above, breaks the current implementation of `Value.Call` and `Value.CallSlice` and a test.

It is unclear to me how to proceed or if in fact I have made a mistake somewhere.

**Comment by ianlancetaylor** [comment]
@nrwiersma What is the status of your work on this issue?  I see that we now have https://go.dev/cl/482375 to implement this.

**Comment by nrwiersma** [comment]
@ianlancetaylor I ran into a road block which I don't know how to resolve without changing the behaviour of `Value.Call` and a test. Details are here: https://github.com/golang/go/issues/49340#issuecomment-1437299990.

I saw the implementation, and it seems it suffers the same issue I ran into, allowing 
```
var i int
out := []Value{ValueOf(i)}
```

**Comment by ianlancetaylor** [comment]
Sorry, I'm not sure I entirely understand what you mean.  Can you give a short example using `Value.Caller` where this causes a problem?  Thanks.

**Comment by nrwiersma** [comment]
Sure. This is from a test I expected to fail, and it does, kind of.
```
fn := func() int {
	return 27
}

var i int
v := ValueOf(i)
out := []Value{v}

ValueOf(fn).Caller().Call([]Value{}, out)

if i != 27 {
	t.Errorf("unexpected result for output value; got %d, want 27", i)
}
```

The actual error you get is `unexpected result for output value; got 27, want 27`. `v` actually should be `v := ValueOf(&i).Elem()`.

The issue here is that `v` is not addressable, and normally this should panic. But as `call` uses flags `flagIndir | flag(tv.Kind())` and to make the output slice reusable, this is allowed and undetectable (I dont know if `call` made the output or if it was provided by the user). If I were to enforce `flagAddr` on the output values I ether cannot reuse output values created by `call` as the dont have this flag or I add `flagAddr` to the `call` created output values and I break the existing behaviour of `Value.Call` and a test (I believe `TestFunc`). That is my conundrum.

**Comment by cuiweixie** [comment]
> Sure. This is from a test I expected to fail, and it does, kind of.
> 
> ```
> fn := func() int {
> 	return 27
> }
> 
> var i int
> v := ValueOf(i)
> out := []Value{v}
> 
> ValueOf(fn).Caller().Call([]Value{}, out)
> 
> if i != 27 {
> 	t.Errorf("unexpected result for output value; got %d, want 27", i)
> }
> ```
> 
> The actual error you get is `unexpected result for output value; got 27, want 27`. `v` actually should be `v := ValueOf(&i).Elem()`.
> 
> The issue here is that `v` is not addressable, and normally this should panic. But as `call` uses flags `flagIndir | flag(tv.Kind())` and to make the output slice reusable, this is allowed and undetectable (I dont know if `call` made the output or if it was provided by the user). If I were to enforce `flagAddr` on the output values I ether cannot reuse output values created by `call` as the dont have this flag or I add `flagAddr` to the `call` created output values and I break the existing behaviour of `Value.Call` and a test (I believe `TestFunc`). That is my conundrum.

It output in my mac m1:
```
unexpected result for output value; got 0, want 27
```
I think this is reasonable.

**Comment by ianlancetaylor** [comment]
@nrwiersma I see, thanks for the example.  I agree that the behavior is not ideal but I think it's OK.  It seems understandable.  And it doesn't seem like a likely mistake.

**Comment by bcmills** [comment]
> In order for this to be useful, I'm responsible for pooling the `reflect.Caller` objects. How do I do that when I'm managing an unbounded number of discrete functions? For each function, I can't just have one pre-prepared `reflect.Caller` since `reflect.Caller` is not concurrent safe. I need a pool of them for each function. It does not scale to have a `sync.Pool` for each of these.

In looking at the concrete CL, it seems that there is still a `sync.Pool` access internally on each invocation, which can lead to substantial cache contention (see #24479) and scalability issues, especially for `Caller` instances wrapping short and very efficient functions.

If the `reflect.Caller` has that sort of nontrivial buffer management, then I don't really see the point of it: it seems like most of its improvements could be made without an API change, by adding a `sync.Map` or two to store the pre-computed information keyed by `rtype` and `methodIndex`. (I suppose there is still an advantage to making the `out` slice an out-parameter, but it's not clear to me that that optimization is worth defining an entire receiver type.)

Moreover: once we promise that `Caller` is safe for concurrent use we cannot reverse that decision, but if we disallow concurrent use for now, we can easily revisit that decision and make it safe later. (Making a concurrency-safe API no longer so is a breaking change, but making a non-concurrency-safe API safe for concurrent use is strictly compatible.)

So I think it is important that the `Call` and `CallSlice` methods of `reflect.Caller` *not* be safe for concurrent use today.

**Comment by bcmills** [comment]
> What's the behavior if the provided `out` has elements that mismatch what the function actually outputs? I propose:
> 
> If the n-th output type is assignable to `out[n].Type()` and `out[n].CanSet`, then the output is directly stored using the equivalent of `out[n].Set`.

That seems too subtle to me: making a small change to a program that was previously allocating new values could cause it to silently switch to overwriting existing values (or vice-versa), the resulting aliasing bugs seem likely to be difficult to diagnose.

Perhaps instead the `Caller` could require that the `out` slice is always prepopulated with values of correct types for which `CanSet` is true, and provide a separate method method to allocate and initialize values as needed?

That method could look like:
```go
// MakeOut returns a slice of Values suitable for passing as the out argument
// to Call or CallSlice.
func (*Caller) MakeOut() []Value
```

**Comment by nrwiersma** [comment]
> So I think it is important that the Call and CallSlice methods of reflect.Caller not be safe for concurrent use today.

This would mean that `Value.Call` and `Value.CallSlice` could not use `Caller`, meaning `Caller` would need to be a entirely new way of function calling, or the `Value` would need its own `sync.Pool` in order to use it.
Also it would make no sense to provide the `out` values to `Call` and `CallSilce` as the `Caller` should be able to entirely manage the output values.

**Comment by bcmills** [comment]
> This would mean that `Value.Call` and `Value.CallSlice` could not use `Caller`

Why is that a problem? Even in https://go.dev/cl/483255/6 a new `Caller` instance is being allocated for each call to `Value.Call` or `Value.CallSlice` and is not being shared across multiple concurrent calls.

> Also it would make no sense to provide the `out` values to `Call` and \[`CallSlice`\] as the `Caller` should be able to entirely manage the output values.

That doesn't follow. It still seems useful to pass in the `out` slice to allow the user of `Caller` to avoid unnecessary allocations, especially if they are reusing the same allocated slice across multiple calls to different functions.

**Comment by nrwiersma** [comment]
Fair enough. If this is the new direction, I don't have the time to completely restart this PR. In that case I would rather close it and let someone else have a go.

**Comment by ianlancetaylor** [comment]
@bcmills Note that the `sync.Pool` is only accessed when there are no output arguments.

Personally I think the current behavior of the output slice is correct: it gets automatically initialized for you, and if you keep passing the same output slice, it doesn't do any additional allocations.  Yes, if you change the output slice in some way it may not behave as you expect.  Don't do that.  Or, we could panic if the element in the output slice is both initialized and the wrong type.
